### PR TITLE
Performance and efficiency improvements in daemon/server mode

### PR DESCRIPTION
**Pull request by dcbw**
When running in server mode we can make a number of performance and efficiency improvements:

1. Use a shared informer to listen for Pod events from the apiserver and grab pod info from that cache rather than doing direct apiserver requests each time. This reduces apiserver load and retry latency, since multus can poll the local cache more frequently than it should do direct apiserver requests.
2. Use a higher QPS on the long-lived kube client since Multus is a performance critical component. Prevents client-side throttling when cluster is under heavy CNI operation, like during mass pod/namespace deletion.
3. Use protobuf instead of JSON in the kube client. More efficient encoding reduces load on the apiserver.
4. Don't bother checking the readiness indicator file every CNI operation in server mode when the ConfigManager is active, since it already watches the readiness indicator and will exit if it gets deleted. Turns out that os.Stat()-ing the file often returns "file not found" (even though it does!) when Multus is running in a container, and that triggers the backoff behavior which adds large latency to CNI operations.

@s1061123 @dougbtv

### CODE REVIEW

File: pkg/types/conf.go [hunk 1]
```diff
@@ -615,6 +615,21 @@
...
 func GetReadinessIndicatorFile(readinessIndicatorFile string) error {
+	// Quick first backoff up to ~1 second
+	backoff := utilwait.Backoff{
+		Steps:    5,
+		Duration: 75 * time.Millisecond,
+		Factor:   2.0,
+		Jitter:   0.1,
+	}
+	if utilwait.ExponentialBackoff(backoff, func() (bool, error) {
+		_, err := os.Stat(readinessIndicatorFile)
```
[the remark this answers is not in the document]
Comment (dcbw) [reply pkg/types/conf.go]: This commit was dropped and the code referenced is no longer in this PR.

File: pkg/k8sclient/k8sclient.go [hunk 2]
```diff
@@ -396,13 +396,14 @@ func TryLoadPodDelegates(pod *v1.Pod, conf *types.NetConf, clientInfo *ClientInf
...
 	if err != nil {
 		return nil, err
 	}
-
-	logging.Debugf("InClusterK8sClient: in cluster config: %+v", config)
-	return NewClientInfo(config)
+	if clientInfo == nil {
+		return nil, fmt.Errorf("failed to create in-cluster kube client")
+	}
+	return clientInfo, err
```
Comment (dougbtv) [inline pkg/k8sclient/k8sclient.go:406]: @s1061123 just want to make sure this approach works with your incoming changes?

**Review by dougbtv — APPROVED** [review]
I'm OK to move forward with this changeset as-is. Overall this looks like great improvements to the thick plugin approach and it's nice to get another take on it.

I would like to get Tomo's take on it as well, especially in consideration of some incoming changes that Tomo has staged.

Thanks for breaking it into commits, let's not squash those when we merge as these are rather logical. Also, thanks on the thorough refactoring for changed parameter counts, unused exports as well as testing.

**Comment by dcbw** [comment]
Fixed up the codecheck comments (missing function doc and an if/else simplification)

**Comment by maiqueb** [comment]
/cc

This is really good. 

Conceptually really on-board w/ this, especially using protobuf instead of JSON, and informers over frying the K8S API.

I'll take a look ASAP.

**Comment by dcbw** [comment]
One thing to note is that static pods don't seem to show up via the informer, but are preset in the apiserver when queried directly. Server-side pod watch filtering doesn't seem to be the issue since this happened even before I added the `env` bits to the daemonset yaml. But given that static pods are fairly rare and usually long-lived, it shouldn't be a problem to fall back to apiserver access for them.

**Comment by maiqueb** [comment]
> One thing to note is that static pods don't seem to show up via the informer, but are preset in the apiserver when queried directly. Server-side pod watch filtering doesn't seem to be the issue since this happened even before I added the `env` bits to the daemonset yaml. But given that static pods are fairly rare and usually long-lived, it shouldn't be a problem to fall back to apiserver access for them.

Good catch. IMO it is perfectly reasonable to fallback to GETting the pod via API if not found in the informer cache.

**Review by dougbtv — APPROVED** [review]
lgtm! thank you!

**Pull request merged**